src/D2L.CodeStyle.UnsafeStaticCounter/Counter.cs

```diff
@@ -0,0 +1,98 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using D2L.CodeStyle.Analyzers.UnsafeStatics;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.MSBuild;
+using Newtonsoft.Json;
+
+namespace D2L.CodeStyle.UnsafeStaticCounter {
+
+	internal sealed class Counter {
+
+		private readonly string _rootDir;
+		private readonly string _outputFile;
+		private readonly ImmutableArray<DiagnosticAnalyzer> _analyzers;
+		private readonly SemaphoreSlim _semaphore;
+
+		public Counter( Options options ) {
+			_analyzers = ImmutableArray.Create<DiagnosticAnalyzer>( new UnsafeStaticsAnalyzer() );
+			_semaphore = new SemaphoreSlim( options.MaxConcurrency );
+			_rootDir = options.RootDir;
+			_outputFile = options.OutputFile;
+		}
+
+		internal async Task<int> Run() {
+			var results = await AnalyzeProjects();
+			WriteOutputFile( results );
+			Console.WriteLine( "done" );
+			return 0;
+		}
+
+		async Task<AnalyzedResults> AnalyzeProjects() {
+			var projectFiles = Directory.EnumerateFiles( _rootDir, "*.csproj", SearchOption.AllDirectories );
+
+			var tasks = projectFiles.Select( AnalyzeProject );
+			var results = await Task.WhenAll( tasks );
+
+			var combinedResult = results
+				.SelectMany( r => r )
+				.ToArray();
+
+			var finalResult = new AnalyzedResults( combinedResult );
+			return finalResult;
+		}
+
+		void WriteOutputFile( AnalyzedResults results ) {
+			var serializer = JsonSerializer.Create( new JsonSerializerSettings {
+				Formatting = Formatting.Indented
+			} );
+			using( var file = File.Open( _outputFile, FileMode.Create ) )
+			using( var stream = new StreamWriter( file ) ) {
+				serializer.Serialize( stream, results );
+			}
+		}
+
+		async Task<AnalyzedStatic[]> AnalyzeProject( string projectFile ) {
+			try {
+				_semaphore.Wait();
+				using( var workspace = MSBuildWorkspace.Create() ) {
+					var proj = await workspace.OpenProjectAsync( projectFile );

```

Comment:
I'll measure and change in a different PR if there's significant differences in time/space.